### DISCUSSION TITLE
Prisma 4.7.1 on Raspberry PI Node.js project

### DISCUSSION THREAD

**Discussion by DavesPlanet** [prisma/prisma#16888] [Q&A]
I got this demo project working on Windows but when I port it to my fully updated Raspberry PI I have issues with Prisma.
https://github.com/remix-run/remix-jokes
Node.JS project running a SQLite database

I have found https://github.com/prisma/prisma/issues/1505

In order to compile prisma I added the software components
direnv 2.25.2-2
protobuf compiler 3.12.4-1

installed rust with
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh

I checked out prisma-engine 4.7.1 and built it with
cargo build --release

I hve the following env variables set in my projects .env and in the current console using export

env | grep PRISMA
PRISMA_INTROSPECTION_ENGINE_BINARY=/home/pi/projects/prisma-engines-4.7.1/target/release/introspection-engine
PRISMA_QUERY_ENGINE_BINARY=/home/pi/projects/prisma-engines-4.7.1/target/release/query-engine
PRISMA_MIGRATION_ENGINE_BINARY=/home/pi/projects/prisma-engines-4.7.1/target/release/migration-engine
PRISMA_FMT_BINARY=/home/pi/projects/prisma-engines-4.7.1/target/release/prisma-fmt

any and all prisma commands result in a similar error, the same error that sent me down this path:

npx prisma generate
> Downloading Prisma engines for Node-API for linux-arm-openssl-1.1.x [                    ] 0%Error: Not Found https://binaries.prisma.sh/all_commits/272861e07ab64f234d3ffc4094e32bd61775599c/linux-arm-openssl-1.1.x/libquery_engine.so.node.gz

in the remix-jokes projecs:
npm start

> start
> remix-serve build
/home/pi/projects/remix-jokes/node_modules/.prisma/client/index.js:3
    throw new Error(
          ^
Error: @prisma/client did not initialize yet. Please run "prisma generate" and try to import it again.
In case this error is unexpected for you, please report it in https://github.com/prisma/prisma/issues
    at new PrismaClient (/home/pi/projects/remix-jokes/node_modules/.prisma/client/index.js:3:11)
    at Object.<anonymous> (/home/pi/projects/remix-jokes/build/index.js:193:6)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/home/pi/projects/remix-jokes/node_modules/@remix-run/serve/dist/cli.js:42:13)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)

my package.json
{
...
  "prisma": {
    "seed": "node --require esbuild-register prisma/seed.ts"
  },
...
  "dependencies": {
...
    "@prisma/client": "^4.7.1",
...
  },
  "devDependencies": {
...
    "prisma": "^4.7.1",
...
  },
...
}

the /prisma folder was generated on windows and contains my dev.db file, so I don't really need to get anything but the dabase engine running.

How can I get a Node.JS project with a SQLite database backed by Prisma running on a Raspberry PI?

Thanks!

**Comment by pantharshit00** [comment]
Hello @DavesPlanet,

Can run `prisma -v` command and see from where it is trying to read the binaries? 

Personally I use the 64bit version of Raspberry Pi OS now on my Pi 4. In that Prisma works out of the box as we ship 64bit ARM binaries. 

In any case, Prisma should not attempt to download binaries if we are providing custom ones so we should resolve that part in any case if that is the actual problem here.

**Reply by DavesPlanet** [reply]
didn't honestly know I was running 32 or that 64 would help, thanks!

All prisma commands result in the same error
 npx prisma -v
> Downloading Prisma engines for Node-API for linux-arm-openssl-1.1.x [                    ] 0%Error: Not Found https://binaries.prisma.sh/all_commits/272861e07ab64f234d3ffc4094e32bd61775599c/linux-arm-openssl-1.1.x/libquery_engine.so.node.gz

folloed the instructions at
https://www.sindastra.de/p/1345/switch-your-raspberry-pi-to-64-bit-kernel-raspbian

 uname -a
Linux raspberrypi 5.15.76-v8+ #1597 SMP PREEMPT Fri Nov 4 12:16:41 GMT 2022 aarch64 GNU/Linux
pi@raspberrypi:~/projects/remix-jokes $ 

so now I seem to be at 64 bit (because it says -v8)?

npx prisma generate
> Downloading Prisma engines for Node-API for linux-arm-openssl-1.1.x [                    ] 0%Error: Not Found https://binaries.prisma.sh/all_commits/272861e07ab64f234d3ffc4094e32bd61775599c/linux-arm-openssl-1.1.x/libquery_engine.so.node.gz

I will delete and re-import all the node modules later and see if that works better this time, no telling what I've mucked up in there
Thanks!

**Reply by DavesPlanet** [reply]
followed the very simple instructions at
https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch/relational-databases-typescript-postgres

npx prisma displays up the usages of prisma

and then...

npx prisma init
> Downloading Prisma engines for Node-API for linux-arm-openssl-1.1.x [                    ] 0%Error: Not Found https://binaries.prisma.sh/all_commits/272861e07ab64f234d3ffc4094e32bd61775599c/linux-arm-openssl-1.1.x/introspection-engine.gz

**Reply by DavesPlanet** [reply]
the instructions I used to make my PI 64 bit seem sketchy, I am going to try to burn a new 64 bit OS onto the PI.

**Comment by DavesPlanet** [comment]
so now many hours later I'm running a full fledged 64 bit OS on my raspberry pi, and I get this

 npx prisma generate
Need to install the following packages:
  prisma@4.7.1
Ok to proceed? (y) y
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
> Downloading Prisma engines for Node-API for linux-arm64-openssl-1.1.x and linux-arm-openssl-1.1.x [                    ] 0%Error: Not Found https://binaries.prisma.sh/all_commits/272861e07ab64f234d3ffc4094e32bd61775599c/linux-arm-openssl-1.1.x/libquery_engine.so.node.gz

**Comment by DavesPlanet** [comment]
well this is different:

npx prisma -v
Environment variables loaded from .env
prisma                  : 4.7.1
@prisma/client          : 4.7.1
Current platform        : linux-arm64-openssl-1.1.x
Query Engine (Node-API) : libquery-engine 272861e07ab64f234d3ffc4094e32bd61775599c (at ../../.npm/_npx/2778af9cee32ff87/node_modules/@prisma/engines/libquery_engine-linux-arm64-openssl-1.1.x.so.node)
Migration Engine        : migration-engine-cli 272861e07ab64f234d3ffc4094e32bd61775599c (at ../../.npm/_npx/2778af9cee32ff87/node_modules/@prisma/engines/migration-engine-linux-arm64-openssl-1.1.x)
Introspection Engine    : introspection-core 272861e07ab64f234d3ffc4094e32bd61775599c (at ../../.npm/_npx/2778af9cee32ff87/node_modules/@prisma/engines/introspection-engine-linux-arm64-openssl-1.1.x)
Format Binary           : prisma-fmt 272861e07ab64f234d3ffc4094e32bd61775599c (at ../../.npm/_npx/2778af9cee32ff87/node_modules/@prisma/engines/prisma-fmt-linux-arm64-openssl-1.1.x)
Format Wasm             : @prisma/prisma-fmt-wasm 4.7.1-1.272861e07ab64f234d3ffc4094e32bd61775599c
Default Engines Hash    : 272861e07ab64f234d3ffc4094e32bd61775599c
Studio                  : 0.477.0

**Reply by DavesPlanet** [reply]
I mean Prisma still doesn't download but that's the first time it's given me version information with
npx prisma -v

but things like
npx prisma db push
and
npx prisma generate
all end in

Downloading Prisma engines for Node-API for linux-arm64-openssl-1.1.x and linux-arm-openssl-1.1.x [                    ] 0%Error: Not Found https://binaries.prisma.sh/all_commits/272861e07ab64f234d3ffc4094e32bd61775599c/linux-arm-openssl-1.1.x/libquery_engine.so.node.gz

**Reply by pantharshit00** [reply]
Can you share your schema? Looks like you are specifying an outdated target. I only need to confirm the generator and datasource section.

**Reply by pantharshit00** [reply]
I am unable to reproduce this on my Pi:
<img width="1531" alt="image" src="https://user-images.githubusercontent.com/22195362/208600704-2f3a6510-79fc-4ab5-983a-c4cb0c3b5008.png">


Here is the schema I am using:

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = env("DATABASE_URL")
}

model Post {
  id        Int      @default(autoincrement()) @id
  createdAt DateTime @default(now())
  title     String
  content   String?
  published Boolean  @default(false)
  author    User     @relation(fields: [authorId], references: [id])
  authorId  Int
}

model Profile {
  id     Int     @default(autoincrement()) @id
  bio    String?
  user   User    @relation(fields: [userId], references: [id])
  userId Int     @unique
}

model User {
  id      Int      @default(autoincrement()) @id
  email   String   @unique
  name    String?
  posts   Post[]
  profile Profile?
}
```


`linux-arm-openssl-1.1.x` which you have in the last output here doesn't exists so if you have that in your schema please remove it.

**Reply by DavesPlanet** [reply]
you are a genius and a lifesaver. I was throwing solutions from around the net at it and someone told me I needed to replace my generator client with this. Thank you so much for your time!

//generator client {
//  provider      = "prisma-client-js"
//  binaryTargets = ["native","linux-arm-openssl-1.1.x"]
//  output        = "./src/generated/db-client"
//}

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
